### PR TITLE
Fixes #14660: fixed synchronous DbCache::set() with the same key

### LINKED ISSUE #14660: Setting the same cache entry twice during the same second makes the 2nd set fail.
When setting the exact same cache entry during the same second, the 2nd `set()` returns false, making it seem like the set has failed, but it didn't. 

The `yii\caching\DbCache` his `setValue()` function first tries an UPDATE to update the cache, and if that fails, it tries an INSERT via the `addValue()` function. This seems fine, except that the UPDATE query returns the number of rows changed, and by default MySQL returns 0 changed rows when the UPDATE is identical to the existing value. I'm not sure what other databases do.
This causes `setValue()` to think the UPDATE has failed and tries a `addValue()` call, what throws an DUPLICATE KEY Exception since the data is already there. This is handled with a `return false` causing the calling functions to think the cache set has failed.

A fix is to set the `PDO::MYSQL_ATTR_FOUND_ROWS` attribute to `true`. This causes MySQL to return the number of found rows instead the number of changed rows in a UPDATE call.

I'm not sure if this is the correct behavior, so a better way could be to change `setValue()` to see the difference between an existing identical cache entry, or a non-existing cache entry.

### Workaround
```php
return [
    ...
    'components' => [
        ...
        'db' => [
            'class' => 'yii\db\Connection',
            'dsn' => 'mysql:host=*hostname*;dbname=*database*',
            'username' => '*username*',
            'password' => '*password*',
            'attributes' => [
                PDO::MYSQL_ATTR_FOUND_ROWS => true,
            ],
        ],
    ],
];
```

### What steps will reproduce the problem?
```php
if (!Yii::$app->cache->set('key', 'value', 86400))) {
  throw new Exception("O my god, the cache set has failed. (1)");
}

if (!Yii::$app->cache->set('key', 'value', 86400))) {
  throw new Exception("O my god, the cache set has failed. (2)");
}
```
Returns:
O my god, the cache set has failed. (2)

```php
if (!Yii::$app->cache->set('key', 'value', 86400))) {
  throw new Exception("O my god, the cache set has failed. (1)");
}

sleep(1); // sleep a second, the expiry value will change

if (!Yii::$app->cache->set('key', 'value', 86400))) {
  throw new Exception("O my god, the cache set has failed. (2)");
}
```
Returns nothing.

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.12
| PHP version      | 7.0
| Operating system | Ubuntu 16.04

### PR DESCRIPTION
**Pull request by rugabarbo**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #14660

### CODE REVIEW

**Comment by rugabarbo** [comment]
I suggest to ignore codeclimate issue. Should not be doing such refactoring because the code will lose its flexibility and clarity.

**Review by samdark — APPROVED** [review]

File: framework/caching/DbCache.php [hunk 1]
```diff
@@ -192,22 +192,21 @@ protected function getValues($keys)
...
-            return $command->execute();
-        });
+                ])->execute();
+            });
 
-        if ($result) {
             $this->gc();
 
             return true;
+        } catch (\Exception $e) {
```
Comment (rob006) [inline framework/caching/DbCache.php]: What is the reason for hiding this exception?

Comment (rugabarbo) [reply framework/caching/DbCache.php]: I implemented abstract method: https://github.com/yiisoft/yii2/blob/master/framework/caching/Cache.php#L400-L410

As you see it must return boolean and not throw any exception.

But this is a good question: should we hide an exception in the case of DB layer failures?

@samdark, @rob006 what do you think about this?

Comment (rob006) [reply framework/caching/DbCache.php]: We could at least log some errors. Same for `addValue()`.

Comment (rugabarbo) [reply framework/caching/DbCache.php]: Like here? https://github.com/yiisoft/yii2/blob/d880148fd09f827aefee2f418294a72a38579bfb/framework/caching/FileCache.php#L156

**Comment by samdark** [comment]
Merged. Thank you!

**Pull request merged**